std-bits/base/src/main/java/org/enso/base/enso_cloud/TransientHTTPResponseCache.java

```diff
@@ -0,0 +1,419 @@
+
+  private static Optional<Long> maxTotalCacheSizeOverrideTestOnly = Optional.empty();
+
+  private static final Map<String, CacheEntry> cache = new HashMap<>();
+  private static final Map<String, ZonedDateTime> lastUsed = new HashMap<>();
+
+  static EnsoHttpResponse makeRequest(
+      URIWithSecrets unresolvedURI,
+      URI resolvedURI,
+      List<Pair<String, String>> resolvedHeaders,
+      RequestMaker requestMaker)
+      throws IOException, InterruptedException, ResponseTooLargeException {
+    var cacheKey = makeHashKey(resolvedURI, resolvedHeaders);
+    if (cache.containsKey(cacheKey)) {
+      return returnCachedResponse(cacheKey);
+    } else {
+      return makeRequestAndCache(
+          unresolvedURI, resolvedURI, resolvedHeaders, cacheKey, requestMaker);
+    }
+  }
+
+  private static EnsoHttpResponse returnCachedResponse(String cacheKey) throws IOException {
+    markCacheEntryUsed(cacheKey);
+    return buildEnsoHttpResponseFromCacheEntry(cache.get(cacheKey));
+  }
+
+  /**
+   * IOExceptions thrown by the HTTP request are propagated; IOExceptions thrown while storing the
+   * data in the cache are caught.
+   *
+   * <p>Only HTTP 200 responses are cached; all others are returned uncached.
+   */
+  private static EnsoHttpResponse makeRequestAndCache(
+      URIWithSecrets unresolvedURI,
+      URI resolvedURI,
+      List<Pair<String, String>> resolvedHeaders,
+      String cacheKey,
+      RequestMaker requestMaker)
+      throws InterruptedException, IOException, ResponseTooLargeException {
+    assert !cache.containsKey(cacheKey);
+
+    var response = requestMaker.run();
+
+    // Don't cache non-200 repsonses.
+    if (response.statusCode() != 200) {
+      return response;
+    }
+
+    // If we get a content-length header, make room for the new file before
+    // downloading. If the response is too large, throw
+    // ResponseTooLargeException.
+    var sizeMaybe = response.headers().firstValue("content-length").map(Long::parseLong);
+    if (sizeMaybe.isPresent()) {
+      long size = sizeMaybe.get();
+      if (size > getMaxFileSize()) {
+        throw new ResponseTooLargeException(
+            size, getMaxFileSize(), unresolvedURI.render().toString());
+      }
+      makeRoomFor(size);
+    }
+
+    try {
+      // Download the response data.
+      int ttl = calculateTTL(response.headers());
+      ZonedDateTime expiry = getNow().plus(Duration.ofSeconds(ttl));
+      int statusCode = response.statusCode();
+      var headers = response.headers();
+      String responseDataPath = downloadResponseData(cacheKey, unresolvedURI, response);
+      long size = new File(responseDataPath).length();
+
+      // Create a cache entry. Remove old files if we are over the cache size
+      // limit.
+      var cacheEntry =
+          new CacheEntry(resolvedURI, headers, responseDataPath, size, statusCode, expiry);
+      cache.put(cacheKey, cacheEntry);
+      markCacheEntryUsed(cacheKey);
```

Comment:
I'm not sure but is the "Remove old files" comment definitely referring to this piece of code? It seems that removing is done just a few lines below.
```suggestion
      // Create a cache entry.
      var cacheEntry =
          new CacheEntry(resolvedURI, headers, responseDataPath, size, statusCode, expiry);
      cache.put(cacheKey, cacheEntry);
      markCacheEntryUsed(cacheKey);
```

Reply:
Removed.